etk-asm/src/asm.rs

```diff
@@ -244,56 +233,102 @@ impl Assembler {
         Self::default()
     }
 
-    /// Indicate that the input sequence is complete. Returns any errors that
-    /// may remain.
-    pub fn finish(self) -> Result<(), Error> {
-        if let Some(undef) = self.pending.front() {
-            return match undef {
-                RawOp::Op(AbstractOp::Macro(invc)) => error::UndeclaredInstructionMacro {
-                    name: invc.name.clone(),
-                }
-                .fail(),
-                RawOp::Op(op) => {
-                    match op
-                        .clone()
-                        .concretize((&self.declared_labels, &self.declared_macros).into())
-                    {
-                        Ok(_) => unreachable!(),
-                        Err(ops::Error::ContextIncomplete {
-                            source: expression::Error::UnknownMacro { name, .. },
-                            ..
-                        }) => error::UndeclaredExpressionMacro { name }.fail(),
-                        Err(ops::Error::ContextIncomplete {
-                            source: expression::Error::UnknownLabel { .. },
-                            ..
-                        }) => {
-                            let labels = op.expr().unwrap().labels(&self.declared_macros).unwrap();
-                            let declared = self.declared_labels.into_keys().collect();
-                            let invoked: HashSet<_> = labels.into_iter().collect();
-                            let missing = invoked
-                                .difference(&declared)
-                                .cloned()
-                                .collect::<Vec<String>>();
-                            error::UndeclaredLabels { labels: missing }.fail()
-                        }
-                        _ => unreachable!(),
-                    }
-                }
-                // bug: if a variable is used when it isn't available, e.g. push1 $size
-                _ => unreachable!(),
-            };
+    /// Create a new `Assembler` with a known length.
+    pub fn new_internal(actual_len: usize) -> Self {
```

Comment:
If this is an internal function, it should be `pub(crate)`.

---

The Rust naming convention is to use something like `with_len` or `with_length` for a constructor that takes a length.

Reply:
I was wrong about the two tests I thought were wrong. I fixed both tests and uploaded the assembler modification, so this function is no longer needed.